include/mumble/Base64.hpp

```diff
@@ -3,30 +3,134 @@
 public:
-	class P;
+	using StrView      = gsl::span< char >;
+	using StrViewConst = gsl::span< const char >;
+
+	static constexpr size_t decode(const BufView out, StrViewConst in) {
+		constexpr std::array< uint8_t, std::numeric_limits< uint8_t >::max() + 1 > table = {
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 64, 63, 52, 53, 54, 55,
+			56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64, 64, 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12,
+			13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 64, 64, 26, 27, 28, 29, 30, 31, 32,
+			33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+		};
+
+		if (in.empty() || in.size() % 4) {
+			return 0;
+		}
+
+		size_t size = in.size() / 4 * 3;
+		if (auto iter = in.rbegin(); *iter == PAD_CHAR) {
+			size -= *++iter == PAD_CHAR ? 2 : 1;
+		}
+
+		if (!out.data()) {
+			return size;
+		}
+
+		if (out.size() < size) {
+			return 0;
+		}
+
+		ByteView data(reinterpret_cast< uint8_t * >(out.data()), out.size());
+
+		while (!in.empty()) {
+			data[0] = (table[in[0]] << 2) + ((table[in[1]] & 0x30) >> 4);
+
+			if (in[2] == '=') {
+				break;
+			}
+
+			data[1] = ((table[in[1]] & 0x0f) << 4) + ((table[in[2]] & 0x3c) >> 2);
+
+			if (in[3] == '=') {
+				break;
+			}
 
-	Base64();
-	virtual ~Base64();
+			data[2] = ((table[in[2]] & 0x03) << 6) + table[in[3]];
 
-	virtual explicit operator bool();
+			in   = in.subspan(4);
+			data = data.subspan(std::min(size_t{ 3 }, data.size()));
+		}
 
-	virtual size_t decode(const BufView out, const BufViewConst in);
-	static size_t encode(const BufView out, const BufViewConst in);
+		return size;
+	}
+
+	static constexpr size_t encode(StrView out, const BufViewConst in) {
+		constexpr std::array< char, 64 > table = { 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
+												   'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
+												   'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
+												   'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
+												   '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+', '/' };
+
+		if (in.empty()) {
+			return 0;
+		}
+
+		const size_t size = static_cast< size_t >(std::ceil(static_cast< double >(in.size()) / 3) * 4);
```

Comment:
`std::ceil` being `constexpr` seems is non-standard for C++ < v23, so we should not use this here.
```suggestion
		// (a + 2) / 3 == std::ceil(static_cast< double >(a) / 3), but constexpr
		const size_t size = 4 * ((in.size() + 2) / 3);
```